File: app/src/main/java/com/trevorwiebe/trackacow/fragments/SettingsFragment.java
package com.trevorwiebe.trackacow.fragments;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;

import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import android.view.LayoutInflater;
import android.view.View;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.dataLoaders.DeleteAllLocalData;
import com.trevorwiebe.trackacow.dataLoaders.InsertAllLocalChangeToCloud;
import com.trevorwiebe.trackacow.utils.Constants;
import com.trevorwiebe.trackacow.utils.Utility;

public class SettingsFragment extends PreferenceFragmentCompat implements InsertAllLocalChangeToCloud.OnAllLocalDbInsertedToCloud {

    private Context mContext;

    private AlertDialog.Builder mBackingUpData;

    private static final String TAG = "SettingsFragment";

    @Override
    public void onCreatePreferences(Bundle bundle, String rootKey) {
        setPreferencesFromResource(R.xml.settings, rootKey);

        Preference versionPref = findPreference("version");
        try {
            if (versionPref != null) {
                versionPref.setSummary(appVersion());
            }
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }

        Preference accountNamePref = findPreference("account_name");
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        if (user != null) {
            String name = user.getDisplayName();
            String email = user.getEmail();
            String accountName = name + ", " + email;
            if (accountNamePref != null) {
                accountNamePref.setSummary(accountName);
            }
        }

        Preference signOutPref = findPreference("sign_out");
        if (signOutPref != null) {
            signOutPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    boolean isThereLocalData = Utility.isThereNewDataToUpload(mContext);
                    if (isThereLocalData) {
                        AlertDialog.Builder thereIsDataDialog = new AlertDialog.Builder(getContext());
                        thereIsDataDialog.setPositiveButton("Backup and sign out", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                Utility.setNewDataToUpload(getContext(), false);
                                mBackingUpData = new AlertDialog.Builder(getContext());
                                View dialogLoadingLayout = LayoutInflater.from(getContext()).inflate(R.layout.dialog_inserting_data_to_cloud, null);
                                mBackingUpData.setView(dialogLoadingLayout);
                                mBackingUpData.show();
                                new InsertAllLocalChangeToCloud(SettingsFragment.this).execute(getContext());
                            }
                        });
                        thereIsDataDialog.setNeutralButton("Sign out anyway", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                signOut();
                            }
                        });
                        thereIsDataDialog.setTitle("Information will be lost");
                        thereIsDataDialog.setMessage("There is information that hasn't been saved to the cloud.  Signing out will delete this data.  Backup to cloud to save you data. You must have an internet connection prior to backing up.");
                        thereIsDataDialog.show();
                    } else {
                        signOut();
                    }
                    return false;
                }
            });
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mContext = context;
    }

    private String appVersion() throws PackageManager.NameNotFoundException {
        PackageInfo pInfo = mContext.getPackageManager().getPackageInfo(mContext.getPackageName(), 0);
        return pInfo.versionName;
    }

    @Override
    public void onAllLocalDbInsertedToCloud(int resultCode) {
        if (resultCode == Constants.SUCCESS) {
            if (mBackingUpData != null) {
                AlertDialog createDialog = mBackingUpData.create();
                createDialog.dismiss();
            }
            signOut();
        } else {
            String message = "";
            switch (resultCode) {
                case Constants.NO_NETWORK_CONNECTION:
                    message = "You do not have a connection, please connect so you can push your data to the cloud.";
                    break;
                case Constants.ERROR_PUSHING_DATA_TO_CLOUD:
                    message = "There was an error pushing your data to the cloud.";
                    break;
                default:
                    message = "An unknown error occurred while pushing your data to the cloud.";
                    break;
            }
            AlertDialog.Builder errorBackingUpDialog = new AlertDialog.Builder(mContext);
            errorBackingUpDialog.setTitle("There was an error backing up your data.");
            errorBackingUpDialog.setMessage(message);
            errorBackingUpDialog.setPositiveButton("Try again", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    new InsertAllLocalChangeToCloud(SettingsFragment.this).execute(getContext());
                }
            });
            errorBackingUpDialog.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {

                }
            });
            errorBackingUpDialog.setNeutralButton("Sign out anyway", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    signOut();
                }
            });
            errorBackingUpDialog.show();
        }
    }

    private void signOut() {
        FirebaseAuth.getInstance().signOut();
        if (getActivity() != null) {
            getActivity().finish();
        }
        new DeleteAllLocalData().execute(getContext());
        Utility.setNewDataToUpload(mContext, false);
        Utility.setLastSync(mContext, 0);
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/fragment_settings/SettingsFragment.kt
package com.trevorwiebe.trackacow.presentation.fragment_settings

import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.preference.CheckBoxPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.trevorwiebe.trackacow.R
import com.trevorwiebe.trackacow.domain.utils.Utility
import com.trevorwiebe.trackacow.domain.utils.Utility.haveNetworkConnection
import com.trevorwiebe.trackacow.domain.utils.Utility.isThereNewDataToUpload
import com.trevorwiebe.trackacow.domain.utils.Utility.setLastSync
import com.trevorwiebe.trackacow.domain.utils.Utility.setNewDataToUpload
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat() {

    private lateinit var mContext: Context

    private val settingsFragmentViewModel: SettingsFragmentViewModel by viewModels()

    override fun onCreatePreferences(bundle: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.settings, rootKey)

        val offlineMode = findPreference<CheckBoxPreference>("off_line_mode")
        offlineMode?.setOnPreferenceChangeListener { _, newValue ->
            val offlineSetting: Boolean = newValue as Boolean
            Utility.setOfflineMode(offlineSetting, mContext)
            true
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.RESUMED) {
                settingsFragmentViewModel.uiState.collect {
                    val firebaseUser = it.firebaseUser
                    val accountNamePref = findPreference<Preference>("account_name")
                    if (firebaseUser != null) {
                        val name = firebaseUser.displayName
                        val email = firebaseUser.email
                        val accountName = "$name, $email"
                        if (accountNamePref != null) {
                            accountNamePref.summary = accountName
                        }
                    }
                }
            }
        }

        val signOutPref = findPreference<Preference>("sign_out")
        if (signOutPref != null) {
            signOutPref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                if (isThereNewDataToUpload(mContext) && haveNetworkConnection(mContext)) {
                    val thereIsDataDialog: AlertDialog.Builder = AlertDialog.Builder(mContext)
                    thereIsDataDialog.setPositiveButton("Backup and sign out") { _, _ ->
                        settingsFragmentViewModel.onEvent(SettingsFragmentEvent.OnUploadCache)
                        signOut()
                    }
                    thereIsDataDialog.setNeutralButton("Sign out anyway") { _, _ ->
                        signOut()
                    }
                    thereIsDataDialog.setMessage("There is information that hasn't been saved to the cloud.  Signing out will delete this data.  Backup to cloud to save your data. You must have an internet connection prior to backing up.")
                    thereIsDataDialog.show()
                } else if (isThereNewDataToUpload(mContext)) {
                    val networkRequired: AlertDialog.Builder = AlertDialog.Builder(mContext)
                    networkRequired.setMessage("There is information that hasn't been saved to the cloud.  Signing out will delete this data.  Please connect to the network and backup to cloud to save your data.")
                    networkRequired.setPositiveButton("Cancel") { _, _ ->

                    }
                    networkRequired.setNeutralButton("Sign out anyway") { _, _ ->
                        signOut()
                    }
                    networkRequired.show()
                } else {
                    signOut()
                }
                false
            }
        }

        val fiveStarReview = findPreference<Preference>("five_star_review")
        if (fiveStarReview != null) {
            fiveStarReview.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                val packageName = "com.trevorwiebe.trackacow"
                try {
                    startActivity(
                        Intent(
                            Intent.ACTION_VIEW,
                            Uri.parse("market://details?id=$packageName")
                        )
                    )
                } catch (e: ActivityNotFoundException) {
                    startActivity(
                        Intent(
                            Intent.ACTION_VIEW,
                            Uri.parse("https://play.google.com/store/apps/details?id=$packageName")
                        )
                    )
                }
                false
            }
        }

        val versionPref = findPreference<Preference>("version")
        try {
            if (versionPref != null) {
                versionPref.summary = appVersion()
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }

    @Throws(PackageManager.NameNotFoundException::class)
    private fun appVersion(): String {
        val pInfo = mContext.packageManager.getPackageInfo(
            mContext.packageName, 0
        )
        return pInfo.versionName
    }

    private fun signOut() {
        settingsFragmentViewModel.onEvent(SettingsFragmentEvent.OnSignOut)
        setNewDataToUpload((mContext), false)
        setLastSync((mContext), 0)
        if (activity != null) {
            requireActivity().finish()
        }
    }
}